### DISCUSSION TITLE
Face recognition real time Xamarin Forms

### DISCUSSION THREAD

**Discussion by maxrodro** [emgucv/emgucv#460] [General]
Hi everyone,

I am developing an app in Xamarin Forms to detect the facial recognition of employees, but I am not able to recognize each employee from the database of faces in a folder on the device, my satisfactory example is putting the same image in the camera of the employee's device, this same employee is not physically annealed by the EigenFaceRecognizer.

This is my code by sections
**//EigenFaceRecognizer  -> Train**

`

	int index = 0;
    PersonsIDs = new List<int>();
	PersonsImages = new List<Mat>();

`_faceCascadeClassifier = new CascadeClassifier(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData) + "/haarcascade/haarcascade_frontalface_alt_tree.xml");`

	String appDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);	
	var files = System.IO.Directory.GetFiles(appDataFolder, "*.png", SearchOption.AllDirectories);
	foreach (var file in files)
	{
		var image = new Image<Gray, byte>(file).Resize(100, 100, Emgu.CV.CvEnum.Inter.Cubic);
		CvInvoke.EqualizeHist(image, image);
		PersonsImages.Add(image.Mat);

		PersonsIDs.Add(index);
		index++;
	}

	_faceRecognizer = new Emgu.CV.Face.EigenFaceRecognizer(PersonsImages.Count, -1);

	if (PersonsImages.Count > 0)
		_faceRecognizer.Train(PersonsImages.ToArray(), PersonsIDs.ToArray());

	Console.WriteLine("Training Done!!!!");`

**Trained face detection method**

` 

           private void FaceRecognition(Mat imageframe)
            {
            System.Drawing.Rectangle[] faces = _faceCascadeClassifier.DetectMultiScale(imageframe, 1.1, 3, new System.Drawing.Size(100, 100));

            foreach (var face in faces)
            {
                var newImageCrop = new Mat(imageframe, face);
                CvInvoke.Resize(newImageCrop, newImageCrop, new System.Drawing.Size(100, 100), 0, 0, Emgu.CV.CvEnum.Inter.Cubic);
                Mat output = new Mat();
                CvInvoke.CvtColor(newImageCrop, output, Emgu.CV.CvEnum.ColorConversion.Bgr2Gray);
                CvInvoke.EqualizeHist(output, output);
              
                var result = _faceRecognizer.Predict(output);
                var label = result.Label;
                if (result.Label != -1)
                {                    
                    CvInvoke.PutText(imageframe, label, new System.Drawing.Point(face.X - 2, face.Y - 2),
                        Emgu.CV.CvEnum.FontFace.HersheyComplex, 1.0, new Bgr(System.Drawing.Color.Orange).MCvScalar);
                    CvInvoke.Rectangle(imageframe, face, new Bgr(System.Drawing.Color.Green).MCvScalar, 2);
                }
                //here results did not found any know faces
                else
                {
                    CvInvoke.PutText(imageframe, "Unknown", new System.Drawing.Point(face.X - 2, face.Y - 2),
                        Emgu.CV.CvEnum.FontFace.HersheyComplex, 1.0, new Bgr(System.Drawing.Color.Orange).MCvScalar);
                    CvInvoke.Rectangle(imageframe, face, new Bgr(System.Drawing.Color.Red).MCvScalar, 2);
                }

            }
        }`

**what am I doing wrong?**

**Comment by emgucv** [comment]
The best reference for Open CV's facial recognition module can be found here:
https://docs.opencv.org/4.5.1/da/d60/tutorial_face_main.html
Please go through the detail on each of the Face Recognizer and understand the benifits and short comings of the individual facial recognizer.

The EigenFace recognizer is probably the least accurate model out of all the available facial recognizers implemented in Open CV. It is really sensitive to brightness, contrast or rotation changes, and even sensitive to tiny misalignments of facial features. 

I would recommend using LBPH over Eigen Face. 

Cascade Classfier is both slower and less accurate than the face detector from DNN module. For simple usage, please see here:
https://github.com/emgucv/emgucv/blob/master/Emgu.CV.Models/Dnn/FaceDetector.cs

Once you located the faces, you should also align the facial features as a way to improve accuracy. Once the facial region is detected, you should detect facial features (center of eyes, center of mouth etc). Then compute the transformation matrix that map the facial features to specific fixed points in the traning/detection image. Once the transformation is applied, you should also remove/blur any background that accidentially get into the picture. 

Hope that helps.
